File: backend/src/routes/files.ts
import { Router } from 'express';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

const router = Router();

// GET /api/files?path=/some/path
router.get('/', (req, res) => {
  const requestedPath = req.query.path as string || '';
  const homeDir = os.homedir();
  const fullPath = path.join(homeDir, requestedPath);

  // Basic security: ensure the path is within home directory
  if (!fullPath.startsWith(homeDir)) {
    return res.status(403).json({ success: false, error: 'Access denied' });
  }

  try {
    const items = fs.readdirSync(fullPath, { withFileTypes: true });
    const result = items.map(item => {
      const itemPath = path.join(fullPath, item.name);
      const stats = fs.statSync(itemPath);
      return {
        name: item.name,
        type: item.isDirectory() ? 'directory' : 'file',
        size: item.isFile() ? stats.size : null,
        modified: stats.mtime,
      };
    });
    res.json({ success: true, data: result, cwd: requestedPath });
  } catch (error) {
    res.status(500).json({ success: false, error: 'Failed to read directory' });
  }
});

export default router;

File: api/files.go
package api

import (
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
)

type FileItem struct {
	Name     string `json:"name"`
	Type     string `json:"type"`
	Size     *int64 `json:"size"`
	Modified string `json:"modified"`
}

func RegisterFilesRoutes(router *gin.RouterGroup) {
	filesBase := os.Getenv("FILES_BASE_PATH")
	if filesBase == "" {
		filesBase, _ = os.UserHomeDir()
	}

	files := router.Group("/files")
	{
		// Serve static files
		files.StaticFS("/serve", gin.Dir(filesBase, false))

		files.GET("/", func(c *gin.Context) {
			requestedPath := c.Query("path")
			fullPath := filepath.Join(filesBase, requestedPath)

			// Clean paths to prevent directory traversal
			fullPath = filepath.Clean(fullPath)
			cleanBase := filepath.Clean(filesBase)

			if !strings.HasPrefix(fullPath, cleanBase) {
				c.JSON(http.StatusForbidden, gin.H{"success": false, "error": "Access denied"})
				return
			}

			entries, err := os.ReadDir(fullPath)
			if err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"success": false, "error": "Failed to read directory"})
				return
			}

			result := []FileItem{}
			for _, entry := range entries {
				info, err := entry.Info()
				if err != nil {
					continue
				}
				
				item := FileItem{
					Name:     entry.Name(),
					Type:     "file",
					Modified: info.ModTime().Format(http.TimeFormat),
				}
				if entry.IsDir() {
					item.Type = "directory"
				} else {
					size := info.Size()
					item.Size = &size
				}
				result = append(result, item)
			}

			c.JSON(http.StatusOK, gin.H{"success": true, "data": result, "cwd": requestedPath})
		})
	}
}
